tests/bluetooth/ctrl_isoal/src/main.c

```diff
@@ -20,1517 +20,7747 @@
-				(sc ? 0 : PDU_ISO_SEG_TIMEOFFSET_SIZE);
-	} else {
-		/* padding packet */
-		pdu_data[1] = 0;
-	}
+	zassert_not_null(pdu_meta, "");
+	zassert_not_null(pdu_meta->meta, "");
+	zassert_not_null(pdu_meta->pdu, "");
 
-	seg_hdr = (struct pdu_iso_sdu_sh *) pdu_meta.pdu->payload;
-	seg_hdr->sc = sc & 0x1;
-	seg_hdr->cmplt = cmplt & 0x1;
-	/* Note, timeoffset only available in first segment of sdu */
-	seg_hdr->length = length + (sc ? 0 : PDU_ISO_SEG_TIMEOFFSET_SIZE);
-	seg_hdr->timeoffset = 0x123456;
-	uint8_t *plp = ((uint8_t *) seg_hdr) + PDU_ISO_SEG_HDR_SIZE +
-			(sc ? 0 : PDU_ISO_SEG_TIMEOFFSET_SIZE);
-	memcpy(plp, dataptr, length);
-	debug_print_pdu();
+	memset(pdu_meta->meta, 0, sizeof(*pdu_meta->meta));
+	memset(pdu_meta->pdu, 0, sizeof(*pdu_meta->pdu));
+
+	pdu_meta->meta->payload_number = payload_number;
+	pdu_meta->meta->timestamp = timestamp;
+	pdu_meta->meta->status = status;
 
-	/* advance pointer to point at next segment header */
-	seg_hdr = (struct pdu_iso_sdu_sh *) (((uint8_t *) seg_hdr) +
-		seg_hdr->length + PDU_ISO_SEG_HDR_SIZE);
+	pdu_meta->pdu->ll_id = llid;
+	pdu_meta->pdu->length = length;
+	memcpy(pdu_meta->pdu->payload, dataptr, length);
+
+	debug_print_rx_pdu(pdu_meta);
 }
 
 /**
- * @brief   Construct Test meta pdu - framed (add segment)
- * @details Based on parameters construct a meta pdu datastructure by
- *          by copying supplied data pointed to by dataptr
- *          Note: construct_pdu_framed_start must be called
- *                on first segment, i.e. before calling this
- *                function.
- *
- * @param cmplt   Set to 1 if this PDU is last PDU of SDU
- * @param dataptr Pointer to source test data
- * @param length  Length of segment raw data to copy, i.e. payload
- *
+ * Insert a new segment in the given PDU
+ * @param[In]     sc          !Start / Continuation bit
+ * @param[In]     cmplt       Complete bit
+ * @param[In]     time_offset Time Offset (us)
+ * @param[In]     dataptr     Pointer to data to fill in segment
+ * @param[In]     length      Length of data
+ * @param[In/Out] pdu_meta    PDU structure including meta information
+ * @return                    PDU data location index
  */
-void construct_pdu_framed_cont(uint8_t cmplt, uint8_t *dataptr, uint8_t length)
+static uint16_t insert_segment(bool sc, bool cmplt, uint32_t time_offset, uint8_t *dataptr,
+				uint8_t length, struct isoal_pdu_rx *pdu_meta)
 {
-	pdu_data[1] += length  + PDU_ISO_SEG_HDR_SIZE; /* Increase total length of PDU */
-	seg_hdr->sc = 1;
-	seg_hdr->cmplt =  cmplt & 0x1;
-	seg_hdr->length = length;
+	struct pdu_iso_sdu_sh seg_hdr;
+	uint16_t pdu_payload_size;
+	uint8_t hdr_write_size;
+	uint16_t pdu_data_loc;
 
-	uint8_t *plp = ((uint8_t *) seg_hdr) + PDU_ISO_SEG_HDR_SIZE;
+	pdu_payload_size = pdu_meta->pdu->length + length + PDU_ISO_SEG_HDR_SIZE +
+			(sc ? 0 : PDU_ISO_SEG_TIMEOFFSET_SIZE);
+	hdr_write_size = PDU_ISO_SEG_HDR_SIZE + (sc ? 0 : PDU_ISO_SEG_TIMEOFFSET_SIZE);
+	memset(&seg_hdr, 0, sizeof(seg_hdr));
```

Comment:
Will not comment more of these: Generally we (void) cast the return value of the stdlib mem* functions in Zephyr

```suggestion
	(void)memset(&seg_hdr, 0, sizeof(seg_hdr));
```